File: interview-preparation/week-1/ruby/02_min_max_sum.rb
#!/bin/ruby

require 'json'
require 'stringio'

#
# Complete the 'miniMaxSum' function below.
#
# The function accepts INTEGER_ARRAY arr as parameter.
#

def miniMaxSum(arr)
  min = max = min_index = 0
  sorted_arr = arr.sort
  max_index = sorted_arr.size - 1
  (1..4).each do
    min = min + sorted_arr[min_index]
    max = max + sorted_arr[max_index]
    min_index += 1
    max_index -= 1
  end
  puts "#{min} #{max}"
end

arr = gets.rstrip.split.map(&:to_i)

miniMaxSum arr

File: interview-preparation/week-1/go/02_min_max_sum.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"sort"
	"strconv"
	"strings"
)

/*
 * Complete the 'miniMaxSum' function below.
 *
 * The function accepts INTEGER_ARRAY arr as parameter.
 */

func miniMaxSum(arr []int32) {
	min, max := int64(0), int64(0)
	minIndex := 0
	maxIndex := len(arr) - 1

	// Convert []int32 to []int64 for sorting
	sortedArr := make([]int64, len(arr))
	for i, num := range arr {
		sortedArr[i] = int64(num)
	}

	// Sort the array in ascending order
	sort.Slice(sortedArr, func(i, j int) bool {
		return sortedArr[i] < sortedArr[j]
	})

	// Calculate the minimum and maximum sums
	for i := 1; i <= 4; i++ {
		min += sortedArr[minIndex]
		max += sortedArr[maxIndex]
		minIndex++
		maxIndex--
	}

	fmt.Println(min, max)
}

func main() {
	reader := bufio.NewReaderSize(os.Stdin, 16*1024*1024)

	arrTemp := strings.Split(strings.TrimSpace(readLine(reader)), " ")

	var arr []int32

	for i := 0; i < 5; i++ {
		arrItemTemp, err := strconv.ParseInt(arrTemp[i], 10, 64)
		checkError(err)
		arrItem := int32(arrItemTemp)
		arr = append(arr, arrItem)
	}

	miniMaxSum(arr)
}

func readLine(reader *bufio.Reader) string {
	str, _, err := reader.ReadLine()
	if err == io.EOF {
		return ""
	}

	return strings.TrimRight(string(str), "\r\n")
}

func checkError(err error) {
	if err != nil {
		panic(err)
	}
}
